packages/framer-motion/src/render/VisualElement.ts

```diff
@@ -0,0 +1,813 @@
+     */
+    blockInitialAnimation: boolean
+
+    /**
+     * A reference to this VisualElement's projection node, used in layout animations.
+     */
+    projection?: IProjectionNode
+
+    /**
+     * A map of all motion values attached to this visual element. Motion
+     * values are source of truth for any given animated value. A motion
+     * value might be provided externally by the component via props.
+     */
+    values = new Map<string, MotionValue>()
+
+    /**
+     * Tracks whether this VisualElement's React component is currently present
+     * within the defined React tree.
+     */
+    isPresent = true
+
+    /**
+     * The AnimationState, this is hydrated by the animation Feature.
+     */
+    animationState?: AnimationState
+
+    /**
+     * The options used to create this VisualElement. The Options type is defined
+     * by the inheriting VisualElement and is passed straight through to the render functions.
+     */
+    private readonly options: Options
+
+    /**
+     * A reference to the latest props provided to the VisualElement's host React component.
+     */
+    private props: MotionProps
+
+    /**
+     * A map of every subscription that binds the provided or generated
+     * motion values onChange listeners to this visual element.
+     */
+    private valueSubscriptions = new Map<string, VoidFunction>()
+
+    /**
+     * A reference to the ReducedMotionConfig passed to the VisualElement's host React component.
+     */
+    private reducedMotionConfig: ReducedMotionConfig | undefined
+
+    /**
+     * On mount, this will be hydrated with a callback to disconnect
+     * this visual element from its parent on unmount.
+     */
+    private removeFromVariantTree: undefined | VoidFunction
+
+    /**
+     * A reference to the previously-provided motion values as returned
+     * from scrapeMotionValuesFromProps. We use the keys in here to determine
+     * if any motion values need to be removed after props are updated.
+     */
+    private prevMotionValues: MotionStyle = {}
+
+    /**
+     * When values are removed from all animation props we need to search
+     * for a fallback value to animate to. These values are tracked in baseTarget.
+     */
+    private baseTarget: ResolvedValues
+
+    /**
+     * Create an object of the values we initially animated from (if initial prop present).
+     */
+    private initialValues: ResolvedValues
+
+    /**
+     * An object containing a SubscriptionManager for each active event.
+     */
+    private events: {
+        [key: string]: SubscriptionManager<any>
+    } = {}
+
+    /**
+     * An object containing an unsubscribe function for each prop event subscription.
+     * For example, every "Update" event can have multiple subscribers via
+     * VisualElement.on(), but only one of those can be defined via the onUpdate prop.
+     */
+    private propEventSubscriptions: {
+        [key: string]: VoidFunction
```

Comment:
Not sure if possible, but maybe?
```suggestion
        [key: keyof typeof propEventHandlers]: VoidFunction
```

Reply:
This gives me 

```
An index signature parameter type cannot be a literal type or generic type. Consider using a mapped object type instead.
```

Reply:
I looked into improving this, but Typescript won't let me.